### DISCUSSION TITLE
Mkdocs macro support in python

### DISCUSSION THREAD

**Discussion by emiliengarreau** [mkdocstrings/mkdocstrings#441] [Q&A] [answered]
I saw that you provide support for mkdocs macro with the crystal handler.
Will it be available in python as well ?
Thanks

**Comment by pawamoy** [comment, marked as answer]
Sure, why not. What's your use case?

**Reply by emiliengarreau** [reply]
My use case would be to generate template of task in yaml.
Example:

```python
class Task:
   """My task
   
   Attributes:
       tata: Lorem Ipsum
       
   Example:
   
   {{ generate_yaml_example( Task ) }}
  """
  tata: int =  5


# in mkdocs_macro.py
 
def define_env(env):
    """Hook function to define environment variables and macros for the documentation.

    Args:
        env: the handler to manipulate the environment
    """

    @env.macro
    def generate_yam_example(name: str):
        if name == "Task":
              return yaml_of_the_task(Task())
 
```

**Reply by pawamoy** [reply]
Not sure to understand. Your example does not show any use for a variable made accessible by the Python handler. When you say that the Crystal handler provides support for MkDocs Macros, what do you mean exactly? This: https://github.com/mkdocstrings/crystal/blob/master/mkdocstrings/handlers/crystal/macros.py?

In any case, I don't think you'll be able to use Jinja markup in your docstrings, because their contents do not pass through the MkDocs event hooks, they are directly rendered with Python-Markdown. So MkDocs Macros will not be able to render them.

**Reply by emiliengarreau** [reply]
Sorry, for the delay.

I was speaking about:
https://mkdocstrings.github.io/crystal/extras.html#support-for-mkdocs-macros-plugin

> In any case, I don't think you'll be able to use Jinja markup in your docstrings, because their contents do not pass through the MkDocs event hooks, they are directly rendered with Python-Markdown. So MkDocs Macros will not be able to render them.

That's what I thought at first but I saw support in mkdocs macro support in crystal documentation. I was intrigued on how you support it. I may have misread the functionalities of the support.

**Reply by pawamoy** [reply]
No problem! I see, thanks for the link. Well we can provide a pluglet as well that would expose the modules collection, but I'd like to see a use-case for it first. Otherwise if you just want to use MkDocs Macros in your docstrings without particularly using attributes from the Python handler... see the next pararaph:

Coming back to what I said previously: I'm not sure Jinja templating in docstrings would be rendered by MkDocs Macros. Looking at its code, the rendering seems to be done in the `on_page_markdown` event, when the docstrings have not yet been injected in the document. Maybe @fralau could confirm :slightly_smiling_face:

**Reply by emiliengarreau** [reply]
Thanks for the answer :+1:

**Reply by fralau** [reply]
Whoah! I hope I am tracking. Yes it is correct that the rendering of macros is done on the `on_page_markdown` event. 

I can see that [mkdocstrings does actions in the `on_env` event](https://github.com/mkdocstrings/mkdocstrings/blob/main/src/mkdocstrings/plugin.py). 

Just to make sure there is no confusion there, [the Jinja2 templating of mkdocs is not the same thing as the Jinja2 templating of mkdocs-macro](https://mkdocs-macros-plugin.readthedocs.io/en/latest/post_production/#trickling-values-into-the-mkdocs-theme). 

- The Jinja2 templating of mkdocs-macros is for rendering macros in markdown pages (`on markdown` event).
- The Jinja2 templating of mkdocs is for rendering HTML templates.

If you really wanted to make black magic, your Plugin (or whatever code) could exploit the rendering engine of mkdocs-macros (method `render()` of the class MacrosPlugin). With that method, you can render any string you please.

**Reply by pawamoy** [reply]
And mkdocstrings also have its own Jinja templating/env :rofl: 
If we somehow gave access to our mkdocstrings Jinja env, I guess users could indeed add mkdocs-macros' `render` method into it, and then override the docstrings templates to use that method. Dark magic as you point it :smile: 
Thanks for commenting @fralau!
